dags/financial_data_dag.py

```diff
@@ -0,0 +1,128 @@
+"""
+Downloads stock market data from S&P 500 companies and inserts it into CrateDB. 
+
+Prerequisites
+-------------
+In CrateDB, the schema to store this data needs to be created once manually.
+See the file setup/financial_data_schema.sql in this repository.
+
+"""
+import datetime
+from airflow import DAG
+from airflow.hooks.postgres_hook import PostgresHook
+from airflow.providers.postgres.operators.postgres import PostgresOperator
+from airflow.operators.python_operator import PythonOperator
+
+import requests
+from bs4 import BeautifulSoup
+import yfinance as yf
+import pandas as pd
+import math
+import json
+
+def get_sp500_ticker_symbols():
+
+    # getting the html code from S&P 500 wikipedia page
+    url = "https://en.wikipedia.org/wiki/List_of_S%26P_500_companies"
+    r_html = requests.get(url,timeout=2.5).text
+    soup = BeautifulSoup(r_html, 'html.parser')
+
+    components_table = soup.find_all(id="constituents")
+    
+    # the first section (index=0) in the components table contains the headers (which are unimportant in this context)
+    # therefore, only data from index=1 on is taken.
+    # moreover, each row in the table is stored in a <tr> division, so we find all of those. The data is stored in an array, where
+    # each element contains information about a S&P 500 company.
+    data_rows = components_table[0].find("tbody").find_all("tr")[1:]
+
+    tickers = []
+    
+    # now the tickers are extracted from the data
+    for row in range(len(data_rows)):
+        stock = list(filter(None,data_rows[row].text.split("\n")))
+        symbol = stock[0]
+        
+        if (symbol.find('.') != -1):
+            symbol = symbol.replace('.', '-')
+            
+        tickers.append(symbol)
+        
+    tickers.sort()
+    
+    return tickers
+
+def download_YFinance_data_function(start_date):
+
+    tickers = get_sp500_ticker_symbols()
+    data = yf.download(tickers, start=start_date)['Adj Close']
+    return data.to_json()
+
+def prepare_data_function(ti):
+    
+    # pulling data (as string)
+    string_data = ti.xcom_pull(task_ids='download_data_task')
+    
+    # transforming to json
+    json_data = json.loads(string_data)
+    
+    # transforming to dataframe for easier manipulation
+    df = pd.DataFrame.from_dict(json_data, orient='index')
+
+    values_array = []
+    
+    for col in range(len(df.columns)):
+        closing_date = df.columns[col]
+        
+        for row in range(len(df.index)):
+            ticker = df.index[row]
+            adj_close = df.iloc[row, col]
+            
+            if (not(adj_close is None or math.isnan(adj_close))):
+                values_array.append("({},\'{}\',{})".format(closing_date, ticker, adj_close))
```

Comment:
There is a much nicer separation here now with the `INSERT` being in another function 👍. The string generated here is still SQL-specific though. For a full separation of concerns, I'd go for an array of dicts here (or something comparable), and then only translate it to SQL syntax in `insert_data_function`. This way, `prepare_data_function` wouldn't be affected if at any point in the future the database-related insert logic would change.